Create functionality for storing data externally

pstaabp opened the pull request:
This provides webwork2 functionality for a PG problem to store and retrieve data
in a database.  With this (see the attached problems), a user can input data that will be saved in the table `COURSENAME_user_set_data`.

[ext_data.tar.gz](https://github.com/openwebwork/webwork2/files/10571609/ext_data.tar.gz)

For example, in problem #1, the student enters a list of numbers and upon submission,
the data is checked as a list of numbers and then stored for a given key. In problem #2,
the list is retrieved, printed and requested to find the mean of the numbers.

There is new functionality to access the table `COURSENAME_user_set_data` to
fetch, put, and delete data as well as a test for the database in `t/user_set_data/user_set_data.t`.
For the test, it uses a couple of settings in `conf/dev/db_settings.conf` or
`conf/dev/db_settings.dist.conf` (this is probably overkill to have this extra file, but wondered
where to put the test database name.)

There is an flag in `$ce->{pg}{options}{externalData}` to turn on/off this functionality. It is
primarily used to skip code to save/load the data. A new setting is now listed in the Optional Modules
section of the Course Configuration.

Some Questions:

1. Names of anything (module/function names).  I'm not sold on the "external data" as a general
  name. I've thought also of "persistant data".

2. In the database, there is new functionality to access the table `COURSENAME_user_set_data`
  The code doesn't use all of the functionality in `DB.pm` file, but I wanted to include other possibilities.

3. We have talked about other ways of extending this idea.  Perhaps have similar functionality for
  `set_data` or `course_data` with broader scopes.  I don't want to try to build those into this PR
  right now, but want to make sure this is solid to be able to handle those options at
  a later time.

In order to get this to function, you will need to update the database for the test course.

This also needs openwebwork/PG#772 to function as well.

Some todos:

[ ] create `store_number`, `store_string`, `store_matrix` or other functions that are used for storing data of particular types.
[ ] extend to storing/retrieving data on the course, user, set or other scope.

drgrice1 commented:
Note that with the changes to the `.gitignore` files if you do what you said on slack the other day, that is
```
git rm -rf --cached .
git add .
```
the files `DATA/uploads/README` and `htdocs/site_info.txt` are still deleted.

There isn't much that can be done about `htdocs/site_info.txt` for this.  You just have to make sure not to actually remove that from git as that is a file we want in the repository.  We just don't want changes to be shown.  This is really a case of a base design choice with that file though.  This should be handled in an entirely different way in the code.  The way it is currently handled is also bad anyway.

However, the reason the `DATA/uploads/README` file is deleted is the same as what you saw with the `conf/dev` files.  All files in `DATA/*` are ignored.  That means the directory `DATA/uploads` is ignored and not looked into.  So the rule `!*README*` does not apply to that file.  To fix this you would need to do the same thing as with `conf/dev` and add
```
!DATA/uploads
DATA/uploads/*
!DATA/uploads/README
```

drgrice1 commented on lib/WeBWorK/Utils/Rendering.pm:
```diff
@@ -146,7 +147,20 @@ sub constructPGOptions ($ce, $user, $set, $problem, $psvn, $formFields, $transla
+	# if external data is requested, fetch it for a user/set
+	if ($ce->{pg}{options}{externalData}) {
+		$options{external_data} = {};
+		my $db = WeBWorK::DB->new($ce->{dbLayout});
```
This should not be done and should not be here at all.  In no way should a new database handle be created here.  There is already a database handle to use.  It is not passed in to this method though, and should not be.  The external data should be passed in via the translation options and obtained from the database before `renderPG` and `constructPGOptions` is called.

pstaabp replied:
I didn't like doing this.  Seemed unnecessary, but couldn't figure out where a database object would be. 

Should the data be retrieved in WeBWorK::ContentGenerator::Problem in the pre_header_initialize sub?

drgrice1 reviewed (changes requested):
I haven't tested this at all yet, but I have observed critical changes that need to be made by looking at the code.

drgrice1 commented on conf/defaults.config:
```diff
@@ -243,6 +243,13 @@ $pg{options}{showCorrectOnRandomize} = 0;
+# allow a course to use external data (store student data in a database)
+$pg{options}{externalData} = 0;
+
+
```
Another thing here.  In addition to adding this setting here, this should be added commented out in localOverrides.conf.

drgrice1 replied:
Actually, I don't think this option should exist.  Just remove it.

pstaabp replied:
Do you mean that this shouldn't be an option that can be set for a course?  I added it to skip the put/fetch db code if it isn't being used.

drgrice1 replied:
Since you did add the capability to change this from the UI, the option does make sense.  Although, I think the help description needs work.  "Allow a course to store/retrieve data from the database for users" is extremely vague as to what the option is really about, and seems to suggest that if the option is turned off the course will not be able to store or retrieve any data for the database.  There is a lot of other unrelated data in the database that the course can still store and retrieve, i.e., all of the data in all of other tables for the course and the rest of the columns in the table that you have added a column to.

pstaabp commented on conf/defaults.config:
```diff
@@ -1750,6 +1757,12 @@ $ConfigValues = [
 			type => 'boolean'
 		},
+		{
+			var => 'pg{options}{externalData}',
+			doc => x('Use external data'),
+			doc2 => x('Allow a course to store/retrieve data from the database for users.'),
+			type => 'boolean'
+		}
```
Shouldn't perltidy fix this?

drgrice1 replied:
perltidy will fix that, but that file isn't perltidied at this point.  The only files that the `run-perltidy` script works on are those that have the extentions `.pl`, `.pm`, or `.t`.

pstaabp replied:
I don't think perltidy touches this file, since it doesn't have a standard perl suffix.

drgrice1 replied:
You can run perltidy on the file directly with `perltidy --pro=.perltidyrc -bext='/' conf/defaults.config`.
Don't do that though, and instead just make the corrections as I suggested.  The conf files have not been perltidied, and doing so would change much more that just those two lines.  I have perltidied the `doc` section of the defaults.config file though.  I did so just on that portion of the file.  So I would like to keep it tidy.  In vim with `coc.nvim` I can do this by selecting that portion and formatting.

drgrice1 replied:
You use `vscode` right?  You should look into using the `Perl::LanguageServer`.  This would probably provide this functionality for you in vscode.  I know that you can use that module with vscode, but I wouldn't know how to set it up.  A google search would probably help here.  I get https://dev.to/perldean/vscode-as-a-perl-ide-3cco for example.

pstaabp commented on lib/WeBWorK/DB.pm:
```diff
@@ -2251,6 +2251,75 @@ sub getAllMergedProblemVersions {
+	*getUserSetDataWhere    = gen_get_records_where("user_set_data");
+}
+
+sub countUserSetData { return scalar shift->listData(@_) }
```
Yes. Correct.  I should also include this in the test suite.

pstaabp commented on lib/WeBWorK/DB/Record/UserSetData.pm:
```diff
@@ -0,0 +1,37 @@
+
+BEGIN {
+	__PACKAGE__->_fields(
+		key_id  => { type => "TINYBLOB NOT NULL", key => 1 },
```
I had started trying to get this working with an AUTOINCREMENT int, and then also using a key field as well, but wanted the uniqueness of `key_id, user_id, set_id` for this table. I switched back to the webwork2 standard way.  So the `key_id` is the `key` that an author uses to store the data.

drgrice1 replied:
I see.  Hmm.  There could be a potential problem with this though.  Authors are going to need to choose unique keys for their data.  If we start accepting these sort of problems into the OPL or even Contrib, and two authors choose the same key (they often just cut and past what we show in documentation and don't change these sort of things), then this could cause issues.

pstaabp commented on lib/WeBWorK/DB/Record/UserSetData.pm:
```diff
@@ -0,0 +1,37 @@
+BEGIN {
+	__PACKAGE__->_fields(
+		key_id  => { type => "TINYBLOB NOT NULL", key => 1 },
+		user_id => { type => "TINYBLOB NOT NULL", key => 1 },
+		set_id  => { type => "TINYBLOB NOT NULL", key => 1 },
```
I'll switch all of the TINYBLOBs over.

pstaabp commented on lib/WeBWorK/Utils/ProblemProcessing.pm:
```diff
@@ -449,4 +453,42 @@ Comment:    $comment
+	my $pg      = $c->{pg};
+	my $problem = $c->{problem};
+
+	return unless $r->{submitAnswers};
```
I saw this in #1885.

drgrice1 commented:
I was thinking about this, and I wonder if a whole new table is really needed.  Since this data is specific to a set, it seems that this would work with a new column in the user set table.  You could save it as JSON and have multiple keys within that stored value.  Or even just drop this key concept and store the data generally.  Is that key concept really even needed?

pstaabp commented:
Updated the code for cleanup and added functionality to add a single real and a string to the DB.  Here are updated example problems. 

[ext_data.tar.gz](https://github.com/openwebwork/webwork2/files/10662711/ext_data.tar.gz)

pstaabp commented:
> I was thinking about this, and I wonder if a whole new table is really needed. Since this data is specific to a set, it seems that this would work with a new column in the user set table. You could save it as JSON and have multiple keys within that stored value. Or even just drop this key concept and store the data generally. Is that key concept really even needed?

I just saw this comment.  I think this is a good idea.  Let me see what I can do with this.

somiaj commented:
> > I was thinking about this, and I wonder if a whole new table is really needed. Since this data is specific to a set, it seems that this would work with a new column in the user set table. You could save it as JSON and have multiple keys within that stored value. Or even just drop this key concept and store the data generally. Is that key concept really even needed?
> 
> I just saw this comment. I think this is a good idea. Let me see what I can do with this.

Could it simplify things to just store a single hash per set. And to allow some sort of global data being stored, make your methods able to read/write the hash from another set. `$hashref = get_set_data(SETID)`, where if `SETID` is not provided, it gets the data from the current set?

drgrice1 commented:
> Could it simplify things to just store a single hash per set. And to allow some sort of global data being stored, make your methods able to read/write the hash from another set. `$hashref = get_set_data(SETID)`, where if `SETID` is not provided, it gets the data from the current set?

What I am suggesting is that a single JSON object be stored (encoded as a string) per set.  You can store a perl hash in the database, so that doesn't make sense.

The point of this pull request is to make per set data.  Making that accessible from another set really is beyond the scope of what is intended, and I think should not be.

Your suggestions aren't quite making sense either.  What do you mean by making the data read/write from another set?  Where is that going to happen?

somiaj commented:
I see I was over simplifying things and thinking that pg could request the data it needed, instead it has to be supplied the data (hence why the scope would be limited to a single set). I could see a use case where a lab could record data in one set, but then use that data in a different set, so being able to access data in different sets might have a use. But that would require a different method of providing that data to pg.

Would there be a way that pg could request the data (vs being supplied it)?

pstaabp commented:
> I see I was over simplifying things and thinking that pg could request the data it needed, instead it has to be supplied the data (hence why the scope would be limited to a single set). I could see a use case where a lab could record data in one set, but then use that data in a different set, so being able to access data in different sets might have a use. But that would require a different method of providing that data to pg.
> 
> Would there be a way that pg could request the data (vs being supplied it)?

This is definitely a webwork2 version of this and I think is the right way to go in that the data can be fetched and the data this is available to the problem as it is being rendered. 

I haven't quite figured out the best way to handle this for webwork3/standalone renderer yet.  The renderer will not have access to the database, so somehow we're going to need to figure out how to pass the data to the renderer.

drgrice1 commented:
@pstaabp:  For webwork3 this will work much the same as it does for webwork2.  It shouldn't be hard to implement.  Webwork3 will handle saving and retrieving the data.  I don't think the standalone renderer will even need any changes.  It already deals with `KEPT_EXTRA_ANSWERS` since that is what MathQuill uses.

drgrice1 commented:
The standalone renderer might need a new paramater for passing the data back in though.  So a minor change there.

pstaabp commented:
Updated the code to drop the separate table to store the data and create a single column in the set_user table.  This column `external_data` is a TEXT which will store data in key/value pairs via a JSON string.  

The database calls are changed to reflect this and the test is updated.  

Recall to test this, you'll need to update the course in the admin course.

drgrice1 reviewed (commented):
It seems that the `lib/WeBWorK/DB/Record/UserSetData.pm` file can be deleted now that a column in `lib/WeBWorK/DB/Record/UserSet.pm` is being used instead.

drgrice1 commented:
The `user_set_data.t` does not seem to be updated to use the column in `UserSet.t` instead of the UserSetData.pm table.

drgrice1 commented on lib/WeBWorK/DB.pm:
```diff
@@ -2251,6 +2253,86 @@ sub getAllMergedProblemVersions {
 	return $self->{problem_version_merged}->get_records_where($where, $order);
 }
 
+# external user set data functions
```
This code seems out of place now.  It should be moved with the rest of the code for handling a `UserSet`.

drgrice1 reviewed (approved):
I think this pull request is in good shape now.

drgrice1 commented on lib/WeBWorK/Utils/ProblemProcessing.pm:
```diff
@@ -453,4 +458,42 @@ Comment:    $comment
+
+		next if $pg->{PG_ANSWERS_HASH}{$ans_name}{rh_ans}{score} == 0;
+
+		# Process the data according to the declared type.
+		if ($type eq 'numeric_list') {
+			$json_data->{$key} =
+				[ map { $_ + 0 } split(/,/, $pg->{PG_ANSWERS_HASH}{$ans_name}{rh_ans}{student_value}) ];
+		} elsif ($type eq 'string') {
+			$json_data->{$key} = '' . $pg->{PG_ANSWERS_HASH}{$ans_name}{rh_ans}{student_value};
+		} elsif ($type eq 'real') {
+			$json_data->{$key} = 0 + $pg->{PG_ANSWERS_HASH}{$ans_name}{rh_ans}{student_value};
+		}
```
Actually, after looking at this more I think that this needs to be changed.  It is really not a good idea to do this type of processing.  Just store the data that the grader accepts, and don't try to process it into a specific type.  In fact, remove the `$type` entirely.

There are lots of possible problems here.   The first is made clear by the faulty `store_number_list` checker in https://github.com/openwebwork/pg/pull/772.  That checker can be fixed, but the problem is that a problem author might use their own custom checker that might have similar issues.

Another problem is that if someone decided they wanted to store a different type of data, then this would need to be extended to add that.  For example, why wouldn't a list of ordered pairs be something valid that one might want to store, or a list of equations?

drgrice1 reviewed (changes requested):
This has some problems that need further attention.

pstaabp commented:
Cleaned up some of the code.  Also I have attached a couple of pg problems to show how to read in strings, numbers and numeric lists. 
[example_ext_data_1.pg.txt](https://github.com/openwebwork/webwork2/files/11028403/example_ext_data_1.pg.txt)
[ext_data2.pg.txt](https://github.com/openwebwork/webwork2/files/11028404/ext_data2.pg.txt)

@drgrice1 One issue I'm having is that the checking answers inside of PGproblemeditor with this (2nd problem above), generates an error.  It's not fetching the external_data--probably a rpc_render thing.  I wonder if you have a quick fix for this or I can dig in.

pstaabp commented:
This error happens in the PGproblemeditor with either "check answers" or "show correct answers".  I just dumped the contents of `$set` at the top of `Rendering::constructPGOptions` and `$set` seems to just contain a few fields that are defined when clicking "check answers" or "show answers".  

When clicking the "Take Action!" to just rerender, I'm getting what appears to be the set from the database.  

The issue related to this is that the field `external_data` is not present in the `$set` when clicking "check answers" or "show answers".   It almost looks like the `FakeSet`, but I thought we got rid of that recently.

drgrice1 commented:
> This error happens in the PGproblemeditor with either "check answers" or "show correct answers". I just dumped the contents of `$set` at the top of `Rendering::constructPGOptions` and `$set` seems to just contain a few fields that are defined when clicking "check answers" or "show answers".
> 
> When clicking the "Take Action!" to just rerender, I'm getting what appears to be the set from the database.
> 
> The issue related to this is that the field `external_data` is not present in the `$set` when clicking "check answers" or "show answers". It almost looks like the `FakeSet`, but I thought we got rid of that recently.

I am not sure what exactly you are doing.  It is working for me with both "check answers" and "show correct answers".

Can you do a zoom meeting now to talk about it?

pstaabp commented:
Sure.  Or I'm on slack right now and can do one through that.

drgrice1 reviewed (approved):
This is looking good now.

pstaabp commented:
I have a set of 4 problems to illustrate the features.  First prob1 must come before prob2 and triangle1 before triangle2 in a set. 
[ext_data_examples.zip](https://github.com/openwebwork/webwork2/files/11062925/ext_data_examples.zip)

- For problems that require pulling data, I have updated the problems to check to ensure that the data exists so there shouldn't be errors and instead reminding the student to go back and do the previous problem first. 
- The problems triangle1 and triangle2, show how to write a custom data type (list of points in 2D) to be saved and then retrieved and parsed.  This might make a good example for the wiki.
